File: examples/node/get_status.ts
/**
 * Get LinuxCNC Status Example
 *
 * Connects to the gRPC server and displays the current machine status.
 * This is the simplest example - a good starting point for understanding the API.
 *
 * Usage:
 *   npx tsx get_status.ts [--host HOST] [--port PORT]
 */

import { program } from "commander";
import {
  LinuxCNCServiceClient,
  GetStatusRequest,
  LinuxCNCStatus,
  Metadata,
  taskModeToJSON,
  taskStateToJSON,
  execStateToJSON,
  interpStateToJSON,
  coolantStateToJSON,
  credentials,
} from "linuxcnc-grpc";

program
  .option("--host <host>", "gRPC server host", "localhost")
  .option("--port <port>", "gRPC server port", "50051")
  .parse();

const opts = program.opts();
const address = `${opts.host}:${opts.port}`;

console.error(`Connecting to ${address}...`);
const client = new LinuxCNCServiceClient(
  address,
  credentials.createInsecure(),
  {
    'grpc.initial_reconnect_backoff_ms': 1000,
    'grpc.max_reconnect_backoff_ms': 5000,
  }
);

function printStatus(status: LinuxCNCStatus): void {
  console.log("=".repeat(60));
  console.log("LinuxCNC Status");
  console.log("=".repeat(60));

  // Task status
  const task = status.task;
  if (!task) {
    console.error("No task status available");
    return;
  }
  console.log("\n[Task]");
  console.log(`  Mode:       ${taskModeToJSON(task.taskMode)}`);
  console.log(`  State:      ${taskStateToJSON(task.taskState)}`);
  console.log(`  Exec State: ${execStateToJSON(task.execState)}`);
  console.log(`  Interp:     ${interpStateToJSON(task.interpState)}`);
  if (task.file) {
    console.log(`  File:       ${task.file}`);
  }

  // Position
  const position = status.position;
  if (!position) {
    console.error("No position status available");
    return;
  }
  const actualPosition = position.actualPosition;
  if (!actualPosition) {
    console.error("No actual position available");
    return;
  }
  const pos = actualPosition;
  console.log("\n[Position]");
  console.log(
    `  X: ${pos.x.toFixed(4).padStart(10)}  Y: ${pos.y.toFixed(4).padStart(10)}  Z: ${pos.z.toFixed(4).padStart(10)}`
  );
  if (pos.a || pos.b || pos.c) {
    console.log(
      `  A: ${pos.a.toFixed(4).padStart(10)}  B: ${pos.b.toFixed(4).padStart(10)}  C: ${pos.c.toFixed(4).padStart(10)}`
    );
  }

  // Trajectory
  const trajectory = status.trajectory;
  if (!trajectory) {
    console.error("No trajectory status available");
    return;
  }
  console.log("\n[Trajectory]");
  console.log(`  Enabled:    ${trajectory.enabled}`);
  console.log(`  Feed Rate:  ${(trajectory.feedrate * 100).toFixed(1)}%`);
  console.log(`  Rapid Rate: ${(trajectory.rapidrate * 100).toFixed(1)}%`);
  console.log(`  Velocity:   ${trajectory.currentVel.toFixed(2)}`);

  // Joints
  console.log("\n[Joints]");
  for (const joint of status.joints) {
    const homed = joint.homed ? "H" : "-";
    const enabled = joint.enabled ? "E" : "-";
    const fault = joint.fault ? "F" : "-";
    console.log(
      `  Joint ${joint.jointNumber}: [${homed}${enabled}${fault}] pos=${joint.input.toFixed(4).padStart(10)}`
    );
  }

  // Spindles
  if (status.spindles.length > 0) {
    console.log("\n[Spindles]");
    for (const spindle of status.spindles) {
      const direction =
        spindle.direction === -1
          ? "REV"
          : spindle.direction === 1
            ? "FWD"
            : "OFF";
      console.log(
        `  Spindle ${spindle.spindleNumber}: ${direction} @ ${spindle.speed.toFixed(0)} RPM`
      );
    }
  }

  // I/O
  const io = status.io;
  if (!io) {
    console.error("No I/O status available");
    return;
  }
  console.log("\n[I/O]");
  console.log(`  E-stop: ${io.estop ? "ACTIVE" : "OK"}`);
  console.log(`  Mist:   ${coolantStateToJSON(io.mist)}`);
  console.log(`  Flood:  ${coolantStateToJSON(io.flood)}`);

  // Active G-codes
  const gcode = status.gcode;
  if (!gcode) {
    console.error("No G-code status available");
    return;
  }
  if (gcode.activeGcodes.length > 0) {
    const gcodes = gcode.activeGcodes
      .filter((g) => g > 0)
      .map((g) => (g % 10 === 0 ? `G${g / 10}` : `G${(g / 10).toFixed(1)}`));
    console.log("\n[Active G-codes]");
    if (gcodes.length > 10) {
      console.log(`  ${gcodes.slice(0, 10).join(" ")}`);
      console.log(`  ${gcodes.slice(10).join(" ")}`);
    } else {
      console.log(`  ${gcodes.join(" ")}`);
    }
  }

  // Errors
  if (status.errors.length > 0) {
    console.log("\n[Errors]");
    for (const err of status.errors) {
      console.log(`  ${err.type}: ${err.message}`);
    }
  }

  console.log();
}

const deadline = new Date(Date.now() + 5000);
client.getStatus(GetStatusRequest.create(), new Metadata(), { deadline }, (err, status) => {
  if (err) {
    console.error(`gRPC error: ${err.code}: ${err.details}`);
    client.close();
    process.exit(1);
  }
  printStatus(status);
  client.close();
});

File: examples/rust/src/bin/get_status.rs
//! Get LinuxCNC Status Example
//!
//! Connects to the gRPC server and displays the current machine status.
//! This is the simplest example - a good starting point for understanding the API.
//!
//! # Usage
//!
//! ```bash
//! cargo run --bin get_status -- --host localhost --port 50051
//! ```

use clap::Parser;
use linuxcnc_grpc::linuxcnc::linux_cnc_service_client::LinuxCncServiceClient;
use linuxcnc_grpc::linuxcnc::{GetStatusRequest, LinuxCncStatus, TaskMode, TaskState};

#[derive(Parser, Debug)]
#[command(name = "get_status")]
#[command(about = "Get current LinuxCNC machine status")]
struct Args {
    /// gRPC server host
    #[arg(long, default_value = "localhost")]
    host: String,

    /// gRPC server port
    #[arg(long, default_value_t = 50051)]
    port: u16,
}

fn format_task_mode(mode: i32) -> &'static str {
    match TaskMode::try_from(mode) {
        Ok(TaskMode::ModeManual) => "MANUAL",
        Ok(TaskMode::ModeAuto) => "AUTO",
        Ok(TaskMode::ModeMdi) => "MDI",
        _ => "UNKNOWN",
    }
}

fn format_task_state(state: i32) -> &'static str {
    match TaskState::try_from(state) {
        Ok(TaskState::StateEstop) => "ESTOP",
        Ok(TaskState::StateEstopReset) => "ESTOP_RESET",
        Ok(TaskState::StateOn) => "ON",
        Ok(TaskState::StateOff) => "OFF",
        _ => "UNKNOWN",
    }
}

fn print_status(status: &LinuxCncStatus) {
    println!("{}", "=".repeat(60));
    println!("LinuxCNC Status");
    println!("{}", "=".repeat(60));

    // Task status
    if let Some(task) = &status.task {
        println!("\n[Task]");
        println!("  Mode:       {}", format_task_mode(task.task_mode));
        println!("  State:      {}", format_task_state(task.task_state));
        println!("  Exec State: {:?}", task.exec_state);
        println!("  Interp:     {:?}", task.interp_state);
        if !task.file.is_empty() {
            println!("  File:       {}", task.file);
        }
    }

    // Position
    if let Some(position) = &status.position {
        if let Some(pos) = &position.actual_position {
            println!("\n[Position]");
            println!("  X: {:10.4}  Y: {:10.4}  Z: {:10.4}", pos.x, pos.y, pos.z);
            if pos.a != 0.0 || pos.b != 0.0 || pos.c != 0.0 {
                println!("  A: {:10.4}  B: {:10.4}  C: {:10.4}", pos.a, pos.b, pos.c);
            }
        }
    }

    // Trajectory
    if let Some(traj) = &status.trajectory {
        println!("\n[Trajectory]");
        println!("  Enabled:    {}", traj.enabled);
        println!("  Feed Rate:  {:.1}%", traj.feedrate * 100.0);
        println!("  Rapid Rate: {:.1}%", traj.rapidrate * 100.0);
        println!("  Velocity:   {:.2}", traj.current_vel);
    }

    // Joints
    if !status.joints.is_empty() {
        println!("\n[Joints]");
        for joint in &status.joints {
            let homed = if joint.homed { "H" } else { "-" };
            let enabled = if joint.enabled { "E" } else { "-" };
            let fault = if joint.fault { "F" } else { "-" };
            println!(
                "  Joint {}: [{}{}{}] pos={:10.4}",
                joint.joint_number, homed, enabled, fault, joint.input
            );
        }
    }

    // Spindles
    if !status.spindles.is_empty() {
        println!("\n[Spindles]");
        for spindle in &status.spindles {
            let direction = match spindle.direction {
                -1 => "REV",
                1 => "FWD",
                _ => "OFF",
            };
            println!(
                "  Spindle {}: {} @ {:.0} RPM",
                spindle.spindle_number, direction, spindle.speed
            );
        }
    }

    // I/O
    if let Some(io) = &status.io {
        println!("\n[I/O]");
        let estop_str = if io.estop { "ACTIVE" } else { "OK" };
        println!("  E-stop: {}", estop_str);
        println!("  Mist:   {:?}", io.mist);
        println!("  Flood:  {:?}", io.flood);
    }

    // Active G-codes
    if let Some(gcode) = &status.gcode {
        if !gcode.active_gcodes.is_empty() {
            println!("\n[Active G-codes]");
            let gcodes: Vec<String> = gcode
                .active_gcodes
                .iter()
                .filter(|&&g| g > 0)
                .map(|&g| {
                    if g % 10 == 0 {
                        format!("G{}", g / 10)
                    } else {
                        format!("G{:.1}", g as f64 / 10.0)
                    }
                })
                .collect();
            if gcodes.len() > 10 {
                println!("  {}", gcodes[..10].join(" "));
                println!("  {}", gcodes[10..].join(" "));
            } else {
                println!("  {}", gcodes.join(" "));
            }
        }
    }

    // Errors
    if !status.errors.is_empty() {
        println!("\n[Errors]");
        for err in &status.errors {
            println!("  {:?}: {}", err.r#type, err.message);
        }
    }

    println!();
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();
    let addr = format!("http://{}:{}", args.host, args.port);

    let mut client = LinuxCncServiceClient::connect(addr).await?;

    let request = tonic::Request::new(GetStatusRequest {});
    let response = client.get_status(request).await?;
    let status = response.into_inner();

    print_status(&status);

    Ok(())
}
